packages/xod-client/src/editor/components/SuggesterContainer.jsx

```diff
@@ -0,0 +1,118 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import CustomScroll from 'react-custom-scroll';
+
+class SuggesterContainer extends React.Component {
+  constructor(props) {
+    super(props);
+    this.scrollRef = null;
+
+    this.moveTimes = 0;
+    this.moveTimeout = null;
+
+    this.state = {
+      scrollPos: undefined,
+    };
+
+    this.setScrollRef = this.setScrollRef.bind(this);
+    this.getScrollPosition = this.getScrollPosition.bind(this);
+    this.updateScrollPosition = this.updateScrollPosition.bind(this);
+
+    this.renderBlocker = this.renderBlocker.bind(this);
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    const newPos = this.getScrollPosition();
+
+    if (newPos === prevState.scrollPos) return;
+    this.updateScrollPosition(newPos);
+  }
+
+  setScrollRef(el) {
+    this.scrollRef = el;
+  }
+
+  getScrollPosition() {
+    if (this.scrollRef) {
+      const contentWrapper = this.scrollRef.refs.contentWrapper;
+
+      const highlighted = contentWrapper.getElementsByClassName('is-highlighted');
+      if (highlighted.length > 0) {
+        const top = highlighted[0].offsetTop;
+        const height = highlighted[0].clientHeight;
+        const containerHeight = contentWrapper.clientHeight;
+        const scrollPos = this.scrollRef.state.scrollPos;
+
+        const isOutsideUp = (top - height <= scrollPos);
+        const isOutsideDown = (top + height >= (scrollPos + containerHeight) - height);
+
+        if (isOutsideDown) {
+          return (top + height + height) - containerHeight;
+        } else if (isOutsideUp) {
+          return top - height;
+        }
+
+        // If its inside container — don't do anything at all.
+        return this.state.scrollPos;
+      }
+
+      // If nothing highlighted
+      return this.state.scrollPos;
+    }
+    // Default value
+    return undefined;
+  }
+
+  updateScrollPosition(pos) {
+    this.setState({
+      scrollPos: pos,
+    });
+  }
+
+  renderBlocker() {
+    return (this.props.blocked) ? (
+      <div
+        className="Suggester-container-blocker"
+        onMouseMoveCapture={(event) => {
+          event.stopPropagation();
+          clearTimeout(this.moveTimeout);
+
+          if (this.moveTimes > 1) {
+            this.props.disableBlocker();
+          }
+
+          this.moveTimes += 1;
+          this.moveTimeout = setTimeout(() => {
+            this.moveTimes = 0;
+          }, 50);
+        }}
+      />
+    ) : null;
```

Comment:
This is a trick. Its intent is not obvious. So please leave a detailed comment describing the rationale and implementation.